### ISSUE TITLE
TLS certificate command to add the authority for Windows doesn't work except in one specific case

### ISSUE THREAD

**Issue by jurchiks** [dunglas/symfony-docker#429]
https://github.com/dunglas/symfony-docker/blob/main/docs/tls.md

Here it says that the command for Windows is:
`docker compose cp caddy:/data/caddy/pki/authorities/local/root.crt %TEMP%/root.crt && certutil -addstore -f "ROOT" %TEMP%/root.crt`

Running this command in plain CMD just opened from the Start Menu gives me the following error: `no configuration file provided: not found`
Running it in plain CMD with admin changes nothing.

However, if I open the Command Prompt in PhpStorm and run the command there, it gets further than the rest:
![image](https://github.com/dunglas/symfony-docker/assets/542788/6414f6dc-b7f4-43c1-83c5-43183ee4ce8c)
But I can't run the Command Prompt as admin in PhpStorm, there is no such option.

Running it in PowerShell complains about the `&&`:
![image](https://github.com/dunglas/symfony-docker/assets/542788/79b260fa-89fa-4509-8395-5c43ff54b967)
Fair enough, whatever.

Running it in Git Bash/MinGW fails at `%TEMP%`:
![image](https://github.com/dunglas/symfony-docker/assets/542788/32d68046-27ed-46dc-815b-fbf5ac24ffe0)

Running it in WSL also fails at the same variable:
![image](https://github.com/dunglas/symfony-docker/assets/542788/d99f86b8-1985-4659-a380-2611d29e4e9f)

even though I have the `TEMP` environment variable defined in CMD:
![image](https://github.com/dunglas/symfony-docker/assets/542788/59890aa3-2243-4be9-a801-85aa4d63cf5d)
But for some reason not in WSL:
![image](https://github.com/dunglas/symfony-docker/assets/542788/d22a0b0d-eaf7-417f-9c5f-8d30a3e11b5d)
Nor in Git Bash:
![image](https://github.com/dunglas/symfony-docker/assets/542788/e0c1d64b-b920-450a-ba9d-43eff3d7fb8c)

At the very least one of those should have worked.

Essentially, the only thing that I have found that worked, is to split the command into two parts, and run the first part (`docker compose cp caddy:/data/caddy/pki/authorities/local/root.crt %TEMP%/root.crt`) in PhpStorm's Command Prompt, then run the second part (`certutil -addstore -f "ROOT" %TEMP%/root.crt`) in plain CMD with admin rights.

Needless to say, it is ridiculous and I would really like to know why that is and what can be done about it.

Edit: I figured it out. You need to open the Command Prompt with admin rights, then `cd` to the project directory, and THEN the command will work. These specifics really should be noted in the article I linked at the top, so that noone else has to go through the same weirdness I did, especially knowing that a lot of devs these days use WSL by default.

Update: although the certificate was installed successfully, the page still does not work:
![image](https://github.com/dunglas/symfony-docker/assets/542788/bf69d5e3-5fc6-4c7b-9aba-4cbc625c1e41)
Restarting the browser didn't help, and the same error was displayed in Firefox and Edge, both of which I hadn't launched in a long time.

In addition to that, the steps described in the StackOverflow article linked in the main README point 4 ("Open https://localhost in your favorite web browser and [accept the auto-generated TLS certificate](https://stackoverflow.com/a/15076602/1352334)") are no longer applicable because current Chrome does not even have a `Show advanced settings` option in the settings. In fact, the author of the StackOverflow reply himself wrote to `click the lock > Certificate Information`, but I don't even have such an option:
![image](https://github.com/dunglas/symfony-docker/assets/542788/cf2792a0-9243-4539-9a1f-db0e427b4dc4)
So basically all of this is fucked and needs to be revisited.

Update 2: apparently it does not want to work with non-localhost IPs... The requests from `127.0.0.5` were not even being logged in the Caddy container. Changing it back to `localhost` started to work, **but** only after I opened the homepage; trying to access the specific API endpoint I tested on before that gave me error 502 for some reason, but the requests were being logged. This is all just so messy, I wanted to change it from `localhost` to a different IP so that I could run multiple containers on port 80 on different IPs (for example, to test the production version locally).

**Comment by maxhelias** [comment]
Hello,

Sorry to tell you this but I see a lot of misunderstanding. You shouldn't execute a command without understanding its meaning first.
To resume
```
docker compose cp caddy:/data/caddy/pki/authorities/local/root.crt %TEMP%/root.crt
```
This line is to get the CRT from the docker container. So we should be in the project dir to execute this cmd.
```
certutil -addstore -f "ROOT" %TEMP%/root.crt
```
This line is to add/trust the CRT to the root authorities. This cmd requires sudo privilege.

So now your issue :
> Running this command in plain CMD just opened from the Start Menu gives me the following error: no configuration file provided: not found
Running it in plain CMD with admin changes nothing.

Does not match the condition of the first cmd

> Running it in PowerShell complains about the &&:
 
PowerShell is not a normal command prompt.

> Running it in Git Bash/MinGW fails at %TEMP%:

Same as above

> Running it in WSL also fails at the same variable:

Same as above, and here we're no longer in a Windows terminal but a Linux one.

>But for some reason not in WSL:
>Nor in Git Bash:

The reason is that they are not a normal prompt and may have a different context.

> At the very least one of those should have worked.

Nope, and I'm not sure that since WSL you can add a CRT to the root authorities.

> Needless to say, it is ridiculous and I would really like to know why that is and what can be done about it.

It's the right thing to do. To solve this problem, stop using windows.

> These specifics really should be noted in the article I linked at the top, so that noone else has to go through the same weirdness I did, especially knowing that a lot of devs these days use WSL by default.

We can't explain all the particularities of each OS. We've already noted all the cmd for each OS. The rest is up to you to know. 

> In addition to that, the steps described in the StackOverflow article linked in the main README point 4 ("Open [localhost](https://localhost/) in your favorite web browser and [accept the auto-generated TLS certificate](https://stackoverflow.com/a/15076602/1352334)") are no longer applicable because current Chrome does not even have a Show advanced settings option in the settings. In fact, the author of the StackOverflow reply himself wrote to click the lock > Certificate Information, but I don't even have such an option:

This means that the site has no certificate at all. And it doesn't, because a certificate is attached to a domain name or an IP. By default it's on localhost unless you define a domain name as explained here : https://github.com/dunglas/symfony-docker/blob/main/docs/build.md#customizing-the-server-name

> So basically all of this is fucked and needs to be revisited.

In fact, everything is good and perfectly explained, thanks to the contributors who donate their free time.

I don't think you need me to answer the last paragraph, you've understood what's wrong.

Pro tips: change the mapping of the port with docker to multi-container not the ip ;)

**Comment by jurchiks** [comment]
> We can't explain all the particularities of each OS.
Literally all you need to write is that this command only works in CMD with Admin access. That is NOT "all particularities".

> This means that the site has no certificate at all.
The caddy container logs say otherwise.

> unless you define a domain name
I have done that.

> stop using windows
Wow, what a magnificent suggestion... Haven't heard that one before. That's not a solution, that's avoiding the problem.
